### PR TITLE
K-23: API authorization

### LINKED ISSUE #23: implement token for kaos init
#### What is the desired change?

Simple security for the cluster can be improved by safeguarding access with a token (placeholder implemented).

#### How will this change improve kaos?

Add a layer of security to ensure a cluster is not fully accessible without a token.

#### Any thoughts on the implementation approach?

User flow would be as follows:
1) System Admin (or Superhero) creates backend with `kaos build ...`
2) UUID is created (token) and added to backend as ENV (via terraform)
3) UUID is written to `./kaos/config` (have to think of immutability)
4) Data Scientist requires valid UUID (token) for all requests -> additional decorator for `token exists`, and add header to all requests
5) backend requires decorator for auth validity

### PR DESCRIPTION
**Pull request by nithishr**
### Description
Closes #23 

### Checklist

- [x] An issue was first created **before** opening this pull request
- [x] The new code follows the kaos [contribution guidelines](CONTRIBUTING.md)
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests to ensure that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: backend/kaos_backend/util/validators.py [hunk 1]
```diff
@@ -159,3 +162,21 @@ def wrapper(*args, **kwargs):
         return function(*args, **kwargs)
 
     return wrapper
+
+
+def validate_request(function):
```
Comment (aalhour) [inline backend/kaos_backend/util/validators.py]: Could you please rename `validate_request` to a name that expresses the intent of this decorator more? Request validation could be anything from Authorization to Schema checking. How about `authenticate` or `auth_required`?

Here are some of the more popular examples I saw out there:
- [`@login_required` from Flask-Security](https://pythonhosted.org/Flask-Security/api.html)
- [`@auth.login_required` from flask-HTTPAuth](https://flask-httpauth.readthedocs.io/en/latest/)
- [`@authenticate`](https://eddmann.com/posts/using-basic-auth-and-decorators-in-pythons-flask/)

Note that they use the word `login` due to maintenance of Cookies and/or Sessions, for your implementation, you might want to call it differently.

File: backend/kaos_backend/util/validators.py [hunk 2]
```diff
@@ -159,3 +162,21 @@ def wrapper(*args, **kwargs):
...
+    """
+    Validate the token in the header
+    """
+
+    def wrapper(*args, **kwargs):
+        token = os.getenv("TOKEN")
+        if 'Token' not in request.headers:
+            raise AuthorizationError("Token not present in the request")
+        req_token = request.headers['Token']
```
Comment (aalhour) [inline backend/kaos_backend/util/validators.py]: Please use the standard [HTTP `Authorization` Header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Authorization).

File: cli/kaos_cli/facades/backend_facade.py [hunk 3]
```diff
@@ -60,7 +60,9 @@ def _set_build_dir(provider, env):
     def build(self, provider, env, local_backend=False, verbose=False):
         dir_build = self._set_build_dir(provider, env)
         build_dir(dir_build)
-        extra_vars = self._get_vars(provider, dir_build)
+        token = uuid.uuid4()
```
Comment (aalhour) [inline cli/kaos_cli/facades/backend_facade.py]: Please rename `token` to `auth_token`.

File: cli/kaos_cli/facades/backend_facade.py [hunk 4]
```diff
@@ -149,8 +151,8 @@ def _parse_config(dir_build):
         return url, kubeconfig
 
     @staticmethod
-    def _get_vars(provider, dir_build):
-        extra_vars = f"--var config_dir={dir_build} "
+    def _get_vars(provider, dir_build, token=None):
+        extra_vars = f"--var config_dir={dir_build} --var token={token}"
```
Comment (aalhour) [inline cli/kaos_cli/facades/backend_facade.py]: Same comment about `token` name.

**Comment by aalhour** [comment]
@nithishr - Sorry for the late reply.

Since the standard Authorization header gets dropped from Ambassador (as per the last time you debugged), can you at least rename the header to `X-Authorization-Token`, it's more of a standard practice to prepend custom headers names with `X-`. Other than that, the PR looks ok to me.

File: cli/kaos_cli/facades/notebook_facade.py [hunk 5]
```diff
@@ -30,12 +30,16 @@ def user(self):
...
+
     def list(self):
         base_url = self.url
         name = self.workspace
 
         # GET /notebook/<name>
-        r = requests.get(f"{base_url}/notebook/{name}")
+        r = requests.get(f"{base_url}/notebook/{name}", headers={"X-Token": self.token})
         if r.status_code >= 300:
             raise NoNotebookError()
```
Comment (DanielMorales9) [inline cli/kaos_cli/facades/notebook_facade.py]: Based on our pair programming session, you can do error handling as follows: 
```        
if 300 <= r.status_code < 400 :
    raise MyCustomError()
elif 400 <= r.status_code < 500:
    err = Error.from_dict(r.json())
    raise RequestError(err.message)
else:
    raise RequestError(r.text)
```

File: cli/kaos_cli/facades/notebook_facade.py [hunk 6]
```diff
@@ -49,7 +53,8 @@ def upload_data_bundle(self, c, **kwargs):
 
         kwargs['user'] = user
 
-        r = requests.post(f"{base_url}/data/{name}/notebook", data=open(c, 'rb').read(), params=kwargs)
+        r = requests.post(f"{base_url}/data/{name}/notebook", data=open(c, 'rb').read(), params=kwargs,
+                          headers={"X-Token": self.token})
 
         if r.status_code >= 300:
```
Comment (DanielMorales9) [inline cli/kaos_cli/facades/notebook_facade.py]: same as above

File: cli/kaos_cli/facades/notebook_facade.py [hunk 7]
```diff
@@ -62,7 +67,8 @@ def upload_source_bundle(self, c, **kwargs):
 
         kwargs['user'] = user
 
-        r = requests.post(f"{base_url}/notebook/{name}", data=open(c, 'rb').read(), params=kwargs)
+        r = requests.post(f"{base_url}/notebook/{name}", data=open(c, 'rb').read(), params=kwargs,
+                          headers={"X-Token": self.token})
 
         if r.status_code >= 300:
```
Comment (DanielMorales9) [inline cli/kaos_cli/facades/notebook_facade.py]: same as above

File: cli/kaos_cli/facades/notebook_facade.py [hunk 8]
```diff
@@ -74,7 +80,8 @@ def deploy(self, **kwargs):
         user = self.user
 
         kwargs['user'] = user
-        r = requests.post(f"{base_url}/notebook/{name}", params=kwargs)
+        r = requests.post(f"{base_url}/notebook/{name}", params=kwargs,
+                          headers={"X-Token": self.token})
 
         if r.status_code >= 300:
```
Comment (DanielMorales9) [inline cli/kaos_cli/facades/notebook_facade.py]: same as above

**Pull request merged**